File: 3490-find-the-maximum-length-of-valid-subsequence-i/3490-find-the-maximum-length-of-valid-subsequence-i.cpp
class Solution {
public:
    int maximumLength(vector<int>& nums) {
        int odd = 0, even = 0;
        for (int num : nums) {
            if (num % 2 == 0) even++;
            else odd++;
        }

        // Try alternating starting with odd
        int alt1 = 0;
        int last = -1; // -1 means start
        for (int num : nums) {
            int p = num % 2;
            if (alt1 == 0 || p != last) {
                alt1++;
                last = p;
            }
        }

        // Try alternating starting with even
        int alt2 = 0;
        last = -1;
        for (int num : nums) {
            int p = num % 2;
            if (alt2 == 0 || p != last) {
                alt2++;
                last = p;
            }
        }

        return max({odd, even, alt1, alt2});
    }
};

File: 3490-find-the-maximum-length-of-valid-subsequence-i/3490-find-the-maximum-length-of-valid-subsequence-i.cs
public class Solution {
    public int MaximumLength(int[] nums) {
        int od = 0;
        int ev = 0;
        foreach(int num in nums){
            if(num%2 == 1)od++;
            else ev++;
        }

        int alt1 = 0; 
        int last1 = -1;
        foreach(int num in nums){
            int p = num % 2;
            if(alt1 == 0 || p != last1){
                alt1++;
                last1 = p;
            }
        }

        int alt2 = 0;
        int last2 = -1;
        foreach(int num in nums){
            int p = num % 2;
            if(alt2 == 0 || p != last2){
                alt2++;
                last2 = p;
            }
        }

        return Math.Max(Math.Max(od,ev),Math.Max(alt1,alt2));
        
    }
}